crates/noirc_frontend/src/hir_def/types.rs

```diff
@@ -107,18 +131,112 @@ impl Type {
         }
     }
 
-    /// True if this type 'matches' another. Matching is more strict
-    /// than subtyping but less strict than Eq. In particular, a type
-    /// matches another iff it is exactly equal to the other type or
-    /// either type is the Error type.
-    pub fn matches(&self, other: &Type) -> bool {
-        self == other || self == &Type::Error || other == &Type::Error
+    #[allow(clippy::result_unit_err)]
+    pub fn try_bind_to_polymorphic_int(&self, var: &TypeVariable) -> Result<(), ()> {
+        let target_id = match &*var.borrow() {
+            TypeBinding::Bound(_) => unreachable!(),
+            TypeBinding::Unbound(id) => *id,
+        };
+
+        match self {
+            Type::FieldElement(_) | Type::Integer(_, _, _) => {
+                *var.borrow_mut() = TypeBinding::Bound(self.clone());
+                Ok(())
+            }
+            Type::PolymorphicInteger(self_var) => {
+                match &*self_var.borrow() {
+                    TypeBinding::Bound(typ) => typ.try_bind_to_polymorphic_int(var),
+                    // Avoid infinitely recursive bindings
+                    TypeBinding::Unbound(id) if *id == target_id => Ok(()),
```

Comment:
I think the only use of the id is for this check and I was wondering, since references are used thanks to the RefCell, we can't simply compare the references?